File: go-probe/gucset.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"

	"github.com/BCBP-SOLUTIONS-FZC-LLC/platform-pgcommon/pkg/domain"
)

type gucsetCase struct {
	Name          string   `json:"name"`
	UserID        string   `json:"user_id"`
	TenantID      string   `json:"tenant_id"`
	TenantRoles   []string `json:"tenant_roles"`
	ExpectedValid bool     `json:"expected_valid"`
	ExpectedError *string  `json:"expected_error"`
}

type gucsetFixture struct {
	Cases []gucsetCase `json:"cases"`
}

type gucsetResult struct {
	Name           string  `json:"name"`
	Valid          bool    `json:"valid"`
	Error          *string `json:"error"`
	MatchesFixture bool    `json:"matches_fixture"`
}

// errorCode maps a validation error to the fixture's language-agnostic error code strings.
func errorCode(err error) *string {
	if err == nil {
		return nil
	}
	var code string
	switch {
	case errors.Is(err, domain.ErrInconsistentGUCSet):
		code = "inconsistent"
	case errors.Is(err, domain.ErrEmptyTenantRole):
		code = "empty_role"
	case errors.Is(err, domain.ErrInvalidTenantRole):
		code = "invalid_role"
	default:
		code = "unknown:" + err.Error()
	}
	return &code
}

func runGUCSetCheck(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("gucset-check requires a path to gucset_cases.json")
	}
	data, err := os.ReadFile(args[0])
	if err != nil {
		return fmt.Errorf("reading fixture: %w", err)
	}
	var fixture gucsetFixture
	if err := json.Unmarshal(data, &fixture); err != nil {
		return fmt.Errorf("parsing fixture: %w", err)
	}

	results := make([]gucsetResult, 0, len(fixture.Cases))
	allMatch := true
	for _, c := range fixture.Cases {
		g := domain.GUCSet{UserID: c.UserID, TenantID: c.TenantID, TenantRoles: c.TenantRoles}
		verr := g.Validate()
		code := errorCode(verr)
		matches := (verr == nil) == c.ExpectedValid
		if code != nil && c.ExpectedError != nil {
			matches = matches && *code == *c.ExpectedError
		} else {
			matches = matches && (code == nil) == (c.ExpectedError == nil)
		}
		if !matches {
			allMatch = false
		}
		results = append(results, gucsetResult{
			Name:           c.Name,
			Valid:          verr == nil,
			Error:          code,
			MatchesFixture: matches,
		})
	}

	return printJSON(map[string]any{
		"language":  "go",
		"check":     "gucset-check",
		"all_match": allMatch,
		"results":   results,
	})
}

File: python-probe/probe/gucset.py
from __future__ import annotations

import json
import sys

from pgcommon import EmptyTenantRoleError, GUCSet, InconsistentGUCSetError, InvalidTenantRoleError

from probe.util import print_json


def _error_code(exc: Exception | None) -> str | None:
    if exc is None:
        return None
    if isinstance(exc, InconsistentGUCSetError):
        return "inconsistent"
    if isinstance(exc, EmptyTenantRoleError):
        return "empty_role"
    if isinstance(exc, InvalidTenantRoleError):
        return "invalid_role"
    return f"unknown:{exc}"


def run(args: list[str]) -> int:
    if not args:
        print("gucset-check requires a path to gucset_cases.json", file=sys.stderr)
        return 2
    with open(args[0]) as f:
        fixture = json.load(f)

    results = []
    all_match = True
    for case in fixture["cases"]:
        guc = GUCSet(
            user_id=case["user_id"],
            tenant_id=case["tenant_id"],
            tenant_roles=case["tenant_roles"],
        )
        error: Exception | None = None
        try:
            guc.validate()
        except (InconsistentGUCSetError, EmptyTenantRoleError, InvalidTenantRoleError) as exc:
            error = exc

        code = _error_code(error)
        valid = error is None
        matches = valid == case["expected_valid"] and code == case["expected_error"]
        if not matches:
            all_match = False
        results.append(
            {"name": case["name"], "valid": valid, "error": code, "matches_fixture": matches}
        )

    print_json(
        {"language": "python", "check": "gucset-check", "all_match": all_match, "results": results}
    )
    return 0
